include/mbedtls/pk.h

```diff
@@ -65,22 +67,34 @@
 #define MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE -0x3980  /**< Unavailable feature, e.g. RSA disabled for RSA key. */
 #define MBEDTLS_ERR_PK_SIG_LEN_MISMATCH    -0x3900  /**< The signature is valid but its length is less than expected. */
 #define MBEDTLS_ERR_PK_HW_ACCEL_FAILED     -0x3880  /**< PK hardware accelerator failed. */
+#define MBEDTLS_ERR_PK_INVALID_SIGNATURE   -0x3800  /**< Invalid signature */
+#define MBEDTLS_ERR_PK_BUFFER_TOO_SMALL    -0x3780  /**< Output buffer too small */
+#define MBEDTLS_ERR_PK_NOT_PERMITTED       -0x3700  /**< Operation not permitted */
+
+/**@}*/
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+/** \name Asymmetric cryptography operation contexts */
+/**@{*/
+
 /**
- * \brief          Public key types
- */
+ * \brief           Asymmetric operation context types
+ *
+ * \note            An opaque key may be an RSA or ECC key or a key of an
+ *                  unrecognized type. Call \c mbedtls_pk_can_do() to check
+ *                  whether a key is of a recognized type. */
 typedef enum {
-    MBEDTLS_PK_NONE=0,
-    MBEDTLS_PK_RSA,
-    MBEDTLS_PK_ECKEY,
-    MBEDTLS_PK_ECKEY_DH,
-    MBEDTLS_PK_ECDSA,
-    MBEDTLS_PK_RSA_ALT,
-    MBEDTLS_PK_RSASSA_PSS,
+    MBEDTLS_PK_NONE=0,          /**< Unused context object */
+    MBEDTLS_PK_RSA,             /**< RSA key pair (normal software implementation) with PKCS#1 v1.5 or PSS context */
+    MBEDTLS_PK_ECKEY,           /**< ECC key pair with ECDSA context */
+    MBEDTLS_PK_ECKEY_DH,        /**< ECC key pair with ECDH context */
+    MBEDTLS_PK_ECDSA,           /**< ECC key pair with ECDSA context */
```

Comment:
This doesn't explain the difference between `ECKEY` and `ECDSA`.

@mpg, @hanno-arm, if you feel capable of explaining the difference, please suggest wording. Otherwise the fact that it isn't explained is a good enough indication of “don't try to understand”.

Reply:
First, I just noticed the description above is wrong: `PK_ECKEY_DH` doesn't hold an ECDH context. It's just the same context as `PK_ECKEY`, which is an EC key pair (and not an ECDSA context, though so far both are the same, with restartable ECC they're different and `PK_ECKEY` does not hold an ECDSA context). The difference between `ECKEY_DH` and `ECKEY` is about permissions: `ECKEY_DH` is not allowed to do do signatures/verifications (so currently only to be imported and exported (potentially to an ECDH context)).

Similarly, the intended difference between `PK_ECKEY` and `PK_ECDSA` is that if the PK module had been extended to handle key exchange, then `PK_ECDSA` would not be allowed to perform ECDH. (Again, I agree that mixing representation type and partial permissions that way was a bad idea, just explaining what the type where supposed to mean.)

I'd suggest:
```
+    MBEDTLS_PK_ECKEY,           /**< Generic ECC key pair */
+    MBEDTLS_PK_ECKEY_DH,        /**< ECC key pair restricted to key exchanges */
+    MBEDTLS_PK_ECDSA,           /**< ECC key pair restricted to signature/verification */
```

Reply:
Done